CMakeLists.txt

```diff
@@ -476,6 +476,13 @@ configure_file(cmake/tcmalloc.h.in
                ${CMAKE_CURRENT_BINARY_DIR}/gperftools/tcmalloc.h
                @ONLY)
 
+# Always include config.h, and include it first.
+if(MSVC)
+   add_compile_options(/FI${CMAKE_CURRENT_BINARY_DIR}/config.h)
```

Comment:
my understanding is we're not doing this part anymore right ?

Reply:
While moving `src/windows/config.h` out of `src/` already sufficiently resolve the ambiguity described in the PR,

There's still a potential conflict if a CMake build is performed on a repository that went through a automake build previously.

In this scenario, we'll have:
1. <CMake build dir>/config.h
2. src/config.h

This is an issue because both GCC and MSVC will prefer the header file relative to the source file which includes it.
That means (2), which is the stale automake generated config file will be consumed for source files directly under `src`, while (2) will be consumed for source files under the descendant directories.

Reply:
The same problem can be observed if the same repo was used for a Linux automake build, followed by a VSProj build.

```
Error	C1189	#error:  Could not determine cache line length - unknown architecture	page_heap_test	D:\Users\LennoxHo\source\gperftools-main\src\base\basictypes.h	379		
Error	C1083	Cannot open include file: 'unistd.h': No such file or directory	page_heap_test	D:\Users\LennoxHo\source\gperftools-main\src\common.cc	38		
Error	C1189	#error:  Could not determine cache line length - unknown architecture	libtcmalloc_minimal	D:\Users\LennoxHo\source\gperftools-main\src\base\basictypes.h	379		
Error	C1083	Cannot open include file: 'unistd.h': No such file or directory	addressmap_unittest	D:\Users\LennoxHo\source\gperftools-main\src\base\logging.h	44		
Error	C1083	Cannot open include file: 'unistd.h': No such file or directory	libtcmalloc_minimal	D:\Users\LennoxHo\source\gperftools-main\src\common.cc	38		
Error	C1189	#error:  Could not determine cache line length - unknown architecture	page_heap_test	D:\Users\LennoxHo\source\gperftools-main\src\base\basictypes.h	379		
Error	C1189	#error:  Could not determine cache line length - unknown architecture	libtcmalloc_minimal	D:\Users\LennoxHo\source\gperftools-main\src\base\basictypes.h	379		
Error	C1189	#error:  Could not determine cache line length - unknown architecture	page_heap_test	D:\Users\LennoxHo\source\gperftools-main\src\base\basictypes.h	379		
Error	C1189	#error:  Could not determine cache line length - unknown architecture	libtcmalloc_minimal	D:\Users\LennoxHo\source\gperftools-main\src\base\basictypes.h	379		
Error	C1083	Cannot open include file: 'unistd.h': No such file or directory	page_heap_test	D:\Users\LennoxHo\source\gperftools-main\src\symbolize.cc	41		
Error	C1083	Cannot open include file: 'unistd.h': No such file or directory	libtcmalloc_minimal	D:\Users\LennoxHo\source\gperftools-main\src\symbolize.cc	41		
Error	C1189	#error:  Could not determine cache line length - unknown architecture	page_heap_test	D:\Users\LennoxHo\source\gperftools-main\src\base\basictypes.h	379		
Error	C1083	Cannot open include file: 'sys/resource.h': No such file or directory	thread_dealloc_unittest	D:\Users\LennoxHo\source\gperftools-main\src\tests\testutil.cc	41		
Error	C1083	Cannot open include file: 'unistd.h': No such file or directory	thread_dealloc_unittest	D:\Users\LennoxHo\source\gperftools-main\src\base\logging.h	44		
Error	C1083	Cannot open include file: 'unistd.h': No such file or directory	realloc_unittest	D:\Users\LennoxHo\source\gperftools-main\src\base\logging.h	44		
Error	C1083	Cannot open include file: 'sys/time.h': No such file or directory	frag_unittest	D:\Users\LennoxHo\source\gperftools-main\src\tests\frag_unittest.cc	40		
```
Deleting the stale `src/config.h` fixes those errors.

Reply:
We shouldn't bother about such corner cases. There is always git clean or similar. And both autotools and cmake support separate build directory builds.

For example, did you know that make distclean removes src/windows/gperftools/tcmalloc.h ? Yeah, one artifact we at some point might consider doing something about.

Reply:
And to re-iterate, IMHO this commit's proposed hardcoding of compiler flags knowledge (/FI and -include) is worse than any risks caused by improbable mixups you're referring to.

Reply:
Yeah, you are probably right.

I have one last idea - would you be open to move the automake generated config.h from `<root>/src/` to another directory? say, `<root>/am/`? This can be another solution to avoiding the autotools/vsproj issue.

Reply:
But I am not sure I am ready yet to have src/windows/gperftools/tcmalloc.h generation to be amputated from autoconf (but on the other hand, already each time I do new release I already have to update more than 1 file manually; so I can be convinced perhaps with right arguments (but they have to be perfect))

Lets keep things as is for now. autotools does support separate directory builds and you're probably only person in the world with any risk of autotools vs msvc mixups.

Reply:
BTW I just looked at the diff between windows and non-windows tcmalloc.h. It differs in aligned new detection, which I am wondering we may consider doing via defined(__cpp_aligned_new) and by struct mallinfo, which availability we may consider detecting via templates magic (e.g. similar to recent getpc.h refactoring).

Reply:
I can take a look at this in the follow up PR along with the HAVE_PTHREAD cleanups